refactor: update customerio_audience to use RudderRecordV2Schema

koladilip opened the pull request:
## Description

This PR updates the customerio_audience destination to use the RudderRecordV2Schema and RudderRecordV2 types.

## Changes

- Add RecordAction enum in rudderEvents.ts for better type safety (renamed from RecordActionType)
- Update RudderRecordV2Schema to make identifiers optional and rudderId required
- Remove z.null() from identifiers type in RudderRecordV2Schema
- Update customerio_audience destination to use RudderRecordV2Schema
- Ensure identifiers remains required for customerio_audience
- Update tests to handle new schema requirements

## Testing

All tests are passing, including the component tests for customerio_audience destination.

Resolve INT-3700

devops-github-rudderstack commented:
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4343/index.html)

devops-github-rudderstack commented:
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4343/index.html)

devops-github-rudderstack commented:
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4343/index.html)

koladilip commented:
I've implemented the CodeRabbit suggestions to improve the code quality:

1. Simplified and hardened the identifier-count check in `CustomerIOMessageSchema`:
   - Replaced the verbose `superRefine` block with a more concise `refine` method
   - Added validation to ensure keys are not empty strings using `z.string().min(1)`
   - Simplified the validation logic to a single check for exactly one identifier
   - Improved the error message to be more consistent

2. Derived the TypeScript type directly from the schema to avoid drift:
   - Used `z.infer<typeof CustomerIOMessageSchema>` instead of manually reconstructing the type with `Omit`
   - This ensures the type definition stays in sync with the schema
   - Removed the now-unused import of `RudderRecordV2`

All tests are passing, including the component tests for customerio_audience destination.

devops-github-rudderstack commented:
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4343/index.html)

vinayteki95 commented on src/v0/destinations/customerio_audience/type.ts:
```diff
@@ -63,33 +60,18 @@ export const CustomerIOConnectionConfigSchema = z
+  // For CustomerIO, identifiers is required and must contain exactly one identifier
+  identifiers: z
+    .record(z.string().min(1), z.union([z.string(), z.number()]))
+    .refine((ids) => Object.keys(ids).length === 1, {
```
Mainly to make it clear what the refine function receives as its argument.
```suggestion
    .refine((identifiersObject) => Object.keys(identifiersObject).length === 1, {
```

koladilip replied:
Done

devops-github-rudderstack commented:
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4343/index.html)

koladilip commented on src/types/rudderEvents.ts:
```diff
@@ -47,6 +47,41 @@ export const RudderRecordV1Schema = z.object({
+
+export const RudderRecordV2Schema = z
+  .object({
+    type: z.literal('record'),
```
Addressed
